test/javascripts/modifiers/track-message-visibility-test.js

```diff
@@ -0,0 +1,36 @@
+import { render, settled } from "@ember/test-helpers";
+import { exists } from "discourse/tests/helpers/qunit-helpers";
+import { setupRenderingTest } from "discourse/tests/helpers/component-test";
+import hbs from "htmlbars-inline-precompile";
+import { module, test } from "qunit";
+
+module(
+  "Discourse Chat | Modifier | track-message-visibility",
+  function (hooks) {
+    setupRenderingTest(hooks);
+
+    test("Marks message as visible when it intersects with the viewport", async function (assert) {
+      const template = hbs`<div {{track-message-visibility}}></div>`;
+
+      await render(template);
+      await settled();
+
+      assert.ok(
+        exists("div[data-visible=true]"),
+        "message is marked as visible"
+      );
+    });
+
+    test("Marks message as visible when it doesn't intersect with the viewport", async function (assert) {
```

Comment:
I can't understand why this one fails. 😢 

 It's super flaky when I run it locally. I tried setting the `display: none;`after rendering it to force the observe to re-run, but it's not working either. Do you have any ideas, @davidtaylorhq?

Reply:
`await settled()` can only wait for things that are pending in the Ember-managed RunLoop. It doesn't have any way of waiting for any pending browser-managed events like the IntersectionObserver callbacks.

I think the easiest solution here is probably the [`waitFor()`](https://github.com/emberjs/ember-test-helpers/blob/master/API.md#waitfor) helper. So, instead of `await settled()`, we would do

```
import { waitFor } from "@ember/test-helpers";

...

await waitFor("div[data-visible=false]");
```

Making that change in both tests seems to solve the flakiness for me on local 🤞